packages/@glimmer/runtime/lib/scanner.ts

```diff
@@ -31,81 +33,135 @@ export default class Scanner {
   }
-  buffer.push([ Ops.FlushElement ]);
-  if (block) {
-    let { statements } = block;
-    for (let i = 0; i < statements.length; i++) {
-      buffer.push(statements[i]);
+
+  scan(): WireFormat.Statement[] {
+    let { statements } = this;
+    this.state = LayoutState.BeforeTopLevel;
+
+    let buffer: WireFormat.Statement[] = [];
+
+    for (let i=0; i<statements.length; i++) {
+      this.processStatement(this.statements[i], buffer);
+    }
+
+    buffer.push([Ops.ClientSideStatement, ClientSide.Ops.DidRenderLayout]);
+
+    return buffer;
+  }
+
+  private processStatement(statement: WireFormat.Statement, buffer: WireFormat.Statement[]) {
+    switch (this.state) {
+      case LayoutState.BeforeTopLevel:
+        this.processBeforeTopLevel(statement, buffer);
+        break;
+
+      case LayoutState.InTopLevel:
+        this.processInTopLevel(statement, buffer);
+        break;
+
+      case LayoutState.AfterFlush:
+        buffer.push(statement);
+        break;
+
+      default:
+        throw unreachable();
     }
   }
-  buffer.push([ Ops.CloseElement ]);
-}
 
-function decorateTopLevelElement(
-  tagName: string,
-  symbols: string[],
-  attrs: WireFormat.Statements.Attribute[],
-  buffer: WireFormat.Statement[]) {
-  let attrsSymbol = symbols.push(ATTRS_BLOCK);
-  buffer.push([Ops.ClientSideStatement, ClientSide.Ops.OpenComponentElement, tagName]);
-  buffer.push([Ops.ClientSideStatement, ClientSide.Ops.DidCreateElement]);
-  buffer.push([Ops.Yield, attrsSymbol, EMPTY_ARRAY]);
-  buffer.push(...attrs);
+  private processBeforeTopLevel(statement: WireFormat.Statement, buffer: WireFormat.Statement[]) {
+    if (WireFormat.Statements.isComponent(statement)) {
+      this.processTopLevelComponent(statement, buffer);
+    } else if (WireFormat.Statements.isOpenElement(statement)) {
+      this.processIsOpenElement(statement, buffer);
+    } else {
+      // Should be whitespace
+      buffer.push(statement);
+    }
+  }
+
+  private processTopLevelComponent(statement: WireFormat.Statements.Component, buffer: WireFormat.Statement[]) {
+    let [, tagName, attrs, , block] = statement;
+
+    if (this.env.hasComponentDefinition(tagName, this.meta) && tagName !== this.componentName) {
+      buffer.push(statement);
+      this.state = LayoutState.AfterFlush;
+      return;
+    }
+
+    assert(!this.env.hasComponentDefinition(tagName, this.meta) || tagName === this.componentName, `Cannot use a component (<${tagName}>) as the top-level element in the layout of <${this.componentName}>`);
+
+    this.state = LayoutState.InTopLevel;
+
+    buffer.push([Ops.ClientSideStatement, ClientSide.Ops.SetComponentAttrs, true]);
+    buffer.push([Ops.ClientSideStatement, ClientSide.Ops.OpenComponentElement, tagName]);
+    buffer.push([Ops.ClientSideStatement, ClientSide.Ops.DidCreateElement]);
+
+    for (let i=0; i<attrs.length; i++) {
+      this.processStatement(attrs[i], buffer);
+    }
+
+    this.processStatement([ Ops.FlushElement ], buffer);
```

Comment:
we might want to extract these into `const` since they just get copied into another array anyway